### PR TITLE
Return deep proposals for GET `/proposals`

### LINKED ISSUE #274: Return deep objects on `/proposals`
The `/proposals` endpoint returns shallow objects.

We want the ability to access deep objects from this endpoint, similar to the behavior of `/proposals/{proposalId}`.

We also might want to make this "deep object" return be the default / singular behavior.

There was some discussion about exploring the use of "related resources" in the base "bundle" object that was discussed in #179, but our initial conversation led us to feel this would be an inconvenient approach since clients would need to reconstruct deep relationships in order to get any value.

There was also some discussion about having the deep object return proposal version values as a dictionary / object instead of an array of key value pairs.  This feature would apply to other endpoints as well so it is captured in #276.

### PR DESCRIPTION
**Pull request by slifty**
This PR introduces the concept of `enrichers` to convert shallow objects of various types into deep objects.

It also applies the enrichers to the `/proposals` endpoint.

Resolves #274

### CODE REVIEW

**Comment by slifty** [comment]
@bickelj @jasonaowen Before I mark this PR as ready I would like to apply the `enricher` model to the `GET /proposal/{id}` endpoint.  That said, I think this code / commit is ready for feedback to ensure this approach is copacetic.

**Comment by slifty** [comment]
We spoke about this after our team meeting, and I wanted to attempt to summarize the conclusions:

1. This implementation is an improvement in some ways to the current implementation (although it's VERY important to note we lose concurrency of those calls).  However...

2. The pattern of "load data, do work, load more data, do more work." is an anti pattern.  It's capped here (one query per type of object), but it's still doing database work in node (e.g. the joins).

SO

What I'm going to do is create another PR that uses a pure SQL approach.  This is an approach that @bickelj implemented long ago in his very first pass on deep data, but since then we learned that tinypg accepts type hints and so we woudln't need to write full blown schemas and type guards to allow for this.

We'll see how that PR looks and then we can pick one to refine-and-then merge!

**Review by jasonaowen — APPROVED** [review]
After [talking](https://chat.opentechstrategies.com/#narrow/stream/75-PDC/topic/deep.20loading.20exploration/near/163086) with @slifty, we've agreed to defer #314 or something like it in favor of this approach, with the intention of revisiting after getting some demo-focused feature work completed.

I was able to run this locally and examine the response, as well as prototype updating the data-viewer to stop making the +N requests. I don't have as full a local database as we have in prod - and it might be time for us to work on building those sync scripts we were talking about writing someday - but so far so good!

File: src/__tests__/proposals.int.test.ts [hunk 1]
```diff
@@ -60,13 +85,37 @@ describe('/proposals', () => {
...
                   externalId: 'proposal-1',
                   applicantId: 1,
                   opportunityId: 1,
                   createdAt: expect.stringMatching(isoTimestampPattern) as string,
+                  versions: [{
+                    id: 1,
+                    proposalId: 1,
+                    version: 1,
+                    applicationFormId: 1,
+                    createdAt: expect.stringMatching(isoTimestampPattern) as string,
```
[the remark this answers is not in the document]
Comment (slifty) [reply src/__tests__/proposals.int.test.ts:101]: Oh that's a really great idea -- this pattern is used like a billion times right now.

File: src/database/enrichers/enrichProposalFieldValues.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    isApplicationFormField,
+  );
+
+  return proposalFieldValues.map((proposalFieldValue) => ({
+    ...proposalFieldValue,
+    applicationFormField: applicationFormFields.find(
+      (applicationFormField) => (
+        applicationFormField.id === proposalFieldValue.applicationFormFieldId
+      ),
+    ),
```
[the remark this answers is not in the document]
Comment (slifty) [reply src/database/enrichers/enrichProposalFieldValues.ts:24]: I like it (sorry I missed writing this in there but here we are)

**Comment by jasonaowen** [comment]
Thanks for cleaning up those little things, @slifty!

I think the docs are not quite done, however. The OpenAPI docs have an example of what data might look like, but the data returned differs somewhat in that the data contains the `applicationFormField` inside each `fieldValue` and the example does not.

That's my last unaddressed concern with this PR, so please go ahead and merge once that's resolved!

**Pull request merged**